File: portal/components/manageDocument/manageDocument.js
function linkManageDocument(context) {
    "use strict";
    console.log("Running linkManageDocument");
    var subjectId = context.subjectId;
    var uploadTriggerControl = document.querySelector("#uploadTrigger");
    var verifyTriggerControl = document.querySelector("#verifyTrigger");
    var uploadControlControl = document.querySelector("#uploadControl");
    var verifyControlControl = document.querySelector("#verifyControl");
    var alertControl = document.querySelector('#alert');
    var descriptionControl = document.querySelector("#" + context.descriptionControlId);

    if(context.hasInitialDescriptionError) {
       context.alertControl.innerHTML = context.initialDescritionError;
       context.alertControl.classList.add("error");
    }
    uploadTriggerControl.addEventListener("change", addFile);
    verifyTriggerControl.addEventListener("click", verify);
    setUploadNeeded(context.initialIsUploadNeeded);

    function setUploadNeeded(uploadNeeded) {
        console.log("setUploadNeeded: " + uploadNeeded);
        if(uploadNeeded) {
            uploadControlControl.hidden = false;
            verifyControlControl.hidden = true;
        } else {
            uploadControlControl.hidden = true;
            verifyControlControl.hidden = false;
        }
        console.log("done setUploadNeeded")
    }

    function addFile(e) {
        var file = e.target.files[0];
        if(!file){
            return
        }
        upload(file);
    }

    function upload(file) {
        var url = "/" + context.updateUrlComponent +  "/" + subjectId;
        console.log("Doing upload with url: " + url);
        var formData = new FormData();
        formData.append("file", file);
        post(url, formData)
            .then(onResponse)
            .catch(onResponse);
    }

    function post(url, data) {
        return axios.post(url, data)
            .then(function (response) {
                return response;
            }).catch(function (error) {
                return error.response;
            });
    }

    function onResponse(response) {
        if(response.status >= 400) {
            showResponse("error", response.data);
        } else {
            var theResponse = response.data;
            var responseClass = "success";
            if(theResponse.IsWarning) {
                responseClass = "error"
            }
            showResponse(responseClass, theResponse.Message);
            descriptionControl.innerHTML = theResponse.Description;
            setUploadNeeded(theResponse.UploadNeeded)
        }
    }

    function showResponse(responseClass, response) {
        alertControl.classList.remove("success");
        alertControl.classList.remove("error");
        alertControl.classList.add(responseClass);
        alertControl.innerHTML = response;
    }

    function verify() {
        var request = {
            Description: descriptionControl.innerHTML
        };
        post("/" + context.verifyUrlComponent + "/" + subjectId, JSON.stringify(request))
            .then(onResponse)
            .catch(onResponse);
    }
}

File: portal/components/manageDocument/manageDocument.go
package manageDocument

import (
	"gitlab.bbinfra.net/3estack/alexandria/portal/util"
	"html/template"
)

var uploadTemplate = `
{{define "upload"}}
    <form enctype="multipart/form-data" class="uploadForm" id="uploadControl" hidden>
        <label class="uploadForm__label" for="uploadTrigger">Upload {{.}} file:</label>
        <input class="uploadForm__input" type="file" name="file" id="uploadTrigger">
    </form>
{{end}}
`
var verifyTemplate = `
{{define "verify"}}
  <table id="verifyControl" hidden>
    <tr>
      <td>Verify {{.}}:</td>
      <td><button type="button" id="verifyTrigger">Verify and refresh</button></td>
    <tr>
  </table>
{{end}}
`

var manageDocumentTemplate = `
{{define "manageDocument"}}
<div class="manageDocument">
  {{template "upload" .SubjectWord}}
  {{template "verify" .SubjectWord}}
  <div class="notification" id="alert"></div>
  <script src="https://unpkg.com/axios/dist/axios.min.js"></script>
  <script src="{{.JsUrl}}"></script>
  <script>
    console.log("Going to initialize the context");
    var context = {
      subjectId: {{.SubjectId}},
      hasInitialDescriptionError: {{.HasInitialDescriptionError}},
      initialDescriptionError: {{.InitialDescriptionError}},
      initialIsUploadNeeded: {{.InitialIsUploadNeeded}},
      descriptionControlId: {{.DescriptionControlId}},
      updateUrlComponent: {{.UpdateUrlComponent}},
      verifyUrlComponent: {{.VerifyUrlComponent}}
    };
    console.log("Initialized the context");
    linkManageDocument(context);
  </script>
</div>
{{end}}
`

func ParseManageDocumentTemplate(name string) *template.Template {
	return util.ParseTemplates(name,
		uploadTemplate, verifyTemplate, manageDocumentTemplate)
}

type ManageDocumentContext struct {
	SubjectId                  string
	HasInitialDescriptionError bool
	InitialDescriptionError    string
	InitialIsUploadNeeded      bool
	JsUrl                      string
	DescriptionControlId       string
	UpdateUrlComponent         string
	VerifyUrlComponent         string
	SubjectWord                string
}

type PortalResponse struct {
	Description  string
	Message      string
	UploadNeeded bool
	IsWarning    bool
}

type PortalRequest struct {
	Description string
}
